File: 0493-reverse-pairs/0493-reverse-pairs.cpp
class Solution {
public:
    int Merge(vector <int> &nums, int low, int mid, int high) {
        int total = 0;
        int j = mid+1;
        for (int i = low; i <= mid; i++) {
            while (j <= high && nums[i] > 2LL*nums[j]) {
                j++;
            }
            total += (j-(mid+1));
        }
        
        vector<int> temp;
        int left = low;
        int right = mid+1;
        while (left <= mid && right <= high) {
            if (nums[left] <= nums[right]){
                temp.push_back(nums[left++]);
            } 
            else{
                temp.push_back(nums[right++]);
            }
        }
        while (left <= mid) {
            temp.push_back(nums[left++]);
        }
        while (right <= high) {
            temp.push_back(nums[right++]);
        }
        for (int i = low; i <= high; i++) {
            nums[i] = temp[i-low];
        }
        return total;
    }
    int MergeSort(vector<int> & nums, int low, int high) {
        if (low >= high) return 0;
        int mid = (low + high)/2;
        int x = MergeSort(nums, low, mid);
        int y = MergeSort(nums, mid+1, high);
        int z = Merge(nums, low, mid, high);
        return x+y+z;
    }
    int reversePairs(vector<int>& nums) {
        return MergeSort(nums, 0, nums.size() - 1);   
    }
};

File: 0493-reverse-pairs/0493-reverse-pairs.py
# reverse pairs count using merge sort
class Solution:
    # merge two sorted section
    def merge(self, arr, left, mid, right):
        i, j = left, mid
        temp = []

        while i < mid and j <= right:
            if arr[i] <= arr[j]:
                temp.append(arr[i])
                i += 1
            else:
                temp.append(arr[j])
                j += 1

        # remaining elements
        temp.extend(arr[i:mid]) 
        temp.extend(arr[j:right+1])

        # copy back to original array
        arr[left: right+1] = temp 

    # count total reverse pairs in a section
    def countPairs(self, arr, left, mid, right):
        count  = 0
        j = mid

        for i in range(left, mid):
            while j <= right and arr[i] > 2 * arr[j]:
                j += 1
            count += (j - mid)

        return count

    # merge sort + count reverse pairs
    def countReversePairs(self, arr, left, right):
        if left >= right:
            return 0

        mid = (left+right)//2
        count = self.countReversePairs(arr, left, mid)
        count += self.countReversePairs(arr, mid+1, right)
        # count before merge
        count += self.countPairs(arr, left, mid+1, right)
        # normal merge
        self.merge(arr, left, mid + 1, right)
        
        return count

    # main function
    def reversePairs(self, nums: List[int]) -> int:
        return self.countReversePairs(nums, 0, len(nums)-1)
        